### DISCUSSION TITLE
Error in getINITModel2

### DISCUSSION THREAD

**Discussion by Flash1593** [SysBioChalmers/Human-GEM#395] [Q&A] [answered]
During the execution of getINITModel2 I'm getting an error:

>> TNBC_GEM=getINITModel2(refModel, tissue, celltype, hpaData, arrayData, metabolomicsData, removeGenes, taskFile, useScoresForTasks, printReport, taskStructure, params, paramsFT);
***Generating model for: TNBC
-Using array data
-Using metabolic tasks

Reference model statistics:
	13069 reactions, 3067 genes
	Mean reaction score: 4.2351
	Mean gene score: 8.2564
	Reactions with positive scores: 54.1817%

Pruned model statistics:
	11880 reactions, 2494 genes
	Mean reaction score: 4.2651
	Mean gene score: 8.0803
	Reactions with positive scores: 54.4697%

Reactions essential for tasks:
	382 reactions, 364 genes
	Mean reaction score: 8.2909
	Mean gene score: 8.156
	Reactions with positive scores: 86.6492%

MILP detected.
Error using regexprep
All cells must be char row vectors.

Error in optimizeProb>renameparams (line 193)
    paramlist = regexprep(paramlist,old{i},new{i});

Error in optimizeProb (line 84)
        prob.csense = renameparams(prob.csense, {'L','G','E'}, {'<','>','='});

Error in runINIT (line 312)
res=optimizeProb(prob,params);

Error in getINITModel2 (line 337)
[~, deletedRxnsInINIT, metProduction] = runINIT(simplifyModel(cModel),rxnScores,metabolomicsData,essentialRxnsForTasks,0,true,false,params);
 
>> 

Please see if I'm doing anything incorrect.

**Comment by JonathanRob** [comment]
Hi @Flash1593, thank you for your patience with this and other issues. This is yet again due to a inconsistency that arose between Human-GEM and RAVEN while adding new features. It should have already been fixed on the `develop` branch of Human-GEM, so can I ask you to try and clone the `develop` branch from Human-GEM (instead of `main`), and see if that solves the problem?

I think Human-GEM is due for a new major release which will merge all the changes/fixes from `develop` into `main`, but it won't be immediate.

**Comment by Flash1593** [comment, marked as answer]
Hi, I tried the Human-GEM from the develop branch. 
First the importYaml function gave this error

> >> ihuman = importYaml('Human-GEM.yml');
> Start importing...
> 	1
> 	2
> 	3
> 	4
> 	5
> 
> importing completed
> follow-up processing...Error using getGenesFromGrRules
> Too many input arguments.
> 
> Error in importYaml (line 271)
> [genes, rxnGeneMat] = getGenesFromGrRules(model.grRules, model.genes);

I removed the model.genes from the code & then the import was successful.

checkTasks is also working fine.

But getINITmodel2 is giving the same exact error -

> MILP detected.
> Error using regexprep
> All cells must be char row vectors.
> 
> Error in optimizeProb>renameparams (line 193)
>     paramlist = regexprep(paramlist,old{i},new{i});
> 
> Error in optimizeProb (line 84)
>         prob.csense = renameparams(prob.csense, {'L','G','E'}, {'<','>','='});
> 
> Error in runINIT (line 312)
> res=optimizeProb(prob,params);
> 
> Error in getINITModel2 (line 340)
> [~, deletedRxnsInINIT, metProduction] = runINIT(simplifyModel(cModel),rxnScores,metabolomicsData,essentialRxnsForTasks,0,true,false,params);
 
The problem is in the regexprep function in optimizeProb code. What can I do to rectify the error?

**Reply by JonathanRob** [reply]
@haowang-bioinfo @edkerk note that the `getGenesFromGrRules` function is present in both RAVEN _and_ Human-GEM repos, and has become out of sync.

@Flash1593 apologies for the many errors, it seems that all of these changes are not playing well together. I will look into the tINIT error - I have seen it before and though it was fixed, but that's clearly not the case.

**Reply by edkerk** [reply]
I'll speed up the introduction of `importYaml` and `exportYaml` in RAVEN, which would resolve this (as it involves the updated `getGenesFromGrRules`.

**Reply by mihai-sysbio** [reply]
@Flash1593 we all appreciate your engagement in following up, and we are thankful for your help in identifying these issues and confirming that they get resolved 🙏🏻

**Reply by haowang-bioinfo** [reply]
@JonathanRob it was figured that the copy/addition of `getGenesFromGrRules` from Human-GEM to RAVEN appears to be a recent [change](https://github.com/SysBioChalmers/RAVEN/commit/ec561123e5201ce35553aebf945091c1a0331a40) happened last month.

@edkerk would be nice to notify this as an issue or somewhere, and more importantly preserve the history of [constributors](https://github.com/SysBioChalmers/Human-GEM/blame/main/code/GPRs/getGenesFromGrRules.m).

**Reply by Flash1593** [reply]
Glad, I was able to help in identifying the issues. I know resolving the code will take some time. Still how long will it take to resolve the issue if I may ask?

Also, meanwhile can I use the older GEM versions as I need to quickly make a model extraction. Does the older version work or it would give the same errors?

**Reply by edkerk** [reply]
A new RAVEN version will be released in a few days.

**Reply by JonathanRob** [reply]
Ok I found the error and have [made a PR](https://github.com/SysBioChalmers/RAVEN/pull/429) that fixes the problem.

**Reply by JonathanRob** [reply]
@Flash1593 in the meantime before the fix is introduced in a new release, you can use [this branch](https://github.com/SysBioChalmers/RAVEN/tree/fix/runINIT_csense) of RAVEN to run tINIT.

**Reply by Flash1593** [reply]
The fix actually worked and tINIT is working fine now. Thanks for the help and support for the problem.

**Reply by JonathanRob** [reply]
Ok great! In the future, feel free to open an [Issue](https://github.com/SysBioChalmers/Human-GEM/issues) if your problem looks like a bug or you're getting errors or something. Not that it makes much of a difference, but might fit better.
